File: server/src/Models/DMToolkitModels/JoinTables/SpellItem.cs
using DMToolkit.API.Models.DMToolkitModels.Entities;
using DMToolkit.API.Models.DMToolkitModels.Items.Bases;

namespace DMToolkit.API.Models.DMToolkitModels.JoinTables;

public class SpellItem : IJoinTable
{
    public required string SpellId { get; set; }
    public Spell Spell { get; set; } = null!;
    public required string ItemId { get; set; }
    public ItemDefinitionBase ItemDefinitionBase { get; set; } = null!;
    public int AmountRequired { get; set; } = 0;
}

File: flutter-client/dm_toolkit/lib/models/dm_toolkit/join_tables/spell_item.dart
import 'package:dm_toolkit/models/dm_toolkit/items/bases/item_definition_base.dart';

class SpellItem {
  final ItemDefinitionBase itemDefinitionBase;
  final int amountRequired;

  SpellItem({
    required this.itemDefinitionBase,
    required this.amountRequired
  });

  factory SpellItem.fromJson(Map<String, dynamic> json, List<ItemDefinitionBase> itemDefinitionBases) {
    try {
      var itemId = json['itemId'] as String;
      var requiredItem = itemDefinitionBases.firstWhere((item) => item.id == itemId);

      return SpellItem(
        itemDefinitionBase: requiredItem, 
        amountRequired: json['amountRequired'] as int
      );
    } catch (e) {
      throw FormatException('SpellItem model is invalid.', e);
    }
  }
}